lib/api/model/model.dart

```diff
@@ -698,3 +710,78 @@ class DmMessage extends Message {
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
 }
+
+enum MessageEditState {
+  none,
+  edited,
+  moved;
+
+  // Code adapted from the shared code: web/shared/src/resolve_topic.ts
+  // Pattern for an arbitrary resolved-topic prefix.
+  // These always begin with the canonical prefix, but can go on longer.
+  // It's designed to remove a weird "✔ ✔✔ " prefix, if present.
+  static RegExp resolvedTopicPrefixRe = RegExp('^✔ [ ✔]*');
+
+  /// Whether two topics are equal, ignoring any resolved-topic prefix.
+  ///
+  /// When a topic is resolved, the clients agree on adding a ✔ prefix to the
+  /// topic string. Topics whose only difference is the ✔ prefix are considered
+  /// the same. This helper can be helpful when checking if a message has been
+  /// moved.
+  static bool areSameTopic(String topic, String prevTopic) {
+    // TODO(#744) Extract this to its own home to support "mark as resolve".
+    topic = topic.replaceFirst(resolvedTopicPrefixRe, '');
+    prevTopic = prevTopic.replaceFirst(resolvedTopicPrefixRe, '');
+
+    return topic == prevTopic;
+  }
+
+  static MessageEditState readFromMessage(Map<dynamic, dynamic> json, String key) {
+    // TODO refactor this into a helper that computes this from the serialized
+    // MessageEditHistory.
+    final editHistory = json['edit_history'] as List<dynamic>?;
+    final lastEditTimestamp = json['last_edit_timestamp'] as int?;
+    if (editHistory == null) {
+      return (lastEditTimestamp != null)
+        ? MessageEditState.edited
+        : MessageEditState.none;
+    }
+
+    // Edit history should never be empty whenever it is present
+    assert(editHistory.isNotEmpty);
+
+    bool hasEditedContent = false;
+    bool hasMoved = false;
+    for (final entry in editHistory) {
+      if (entry['prev_content'] != null) {
+        hasEditedContent = true;
+      }
+
+      if (entry['prev_stream'] != null) {
+        hasMoved = true;
+      }
```

Comment:
```suggestion
      if (entry['prev_stream'] != null) {
        hasMoved = true;
        continue;
      }
```

That way if it was moved between channels, we skip having to examine any topic change in the same entry.